lib/Common/Memory/HeapBucket.cpp

```diff
@@ -503,10 +504,42 @@ HeapBucketT<TBlockType>::SnailAlloc(Recycler * recycler, TBlockAllocatorType * a
     AllocationVerboseTrace(recycler->GetRecyclerFlagsTable(), _u("TryAlloc failed, forced collection on allocation [Collected: %d]\n"), collected);
     if (!collected)
     {
+        // wait for background sweeping finish if there are too many pages allocated during background sweeping
+        BOOL isBackgroundSweeping = recycler->IsConcurrentSweepExecutingState();
+        if (newAssignedPageCountDuringBGSweeping > (uint)CONFIG_FLAG(NewPagesCapDuringBGSweeping))
+        {
+            if (isBackgroundSweeping)
+            {
+                recycler->FinishConcurrent<ForceFinishCollection>();
+                memBlock = this->TryAlloc(recycler, allocator, sizeCat, attributes);
+                if (memBlock != nullptr)
+                {
+                    newAssignedPageCountDuringBGSweeping = 0;
+                    return memBlock;
+                }
+            }
+
+            recycler->CollectNow<CollectNowForceInThread>();
+            memBlock = this->TryAlloc(recycler, allocator, sizeCat, attributes);
+            if (memBlock != nullptr)
+            {
+                newAssignedPageCountDuringBGSweeping = 0;
+                return memBlock;
+            }
+        }
+
         // We didn't collect, try to add a new heap block
         memBlock = TryAllocFromNewHeapBlock(recycler, allocator, sizeCat, attributes);
         if (memBlock != nullptr)
         {
+            if (isBackgroundSweeping)
+            {
+                newAssignedPageCountDuringBGSweeping += TBlockType::HeapBlockAttributes::PageCount;
```

Comment:
We can just always increment this count (inside TryAllocFromNewHeapBlock)
And 0 it out before we Sweep.